File: tools/binding_compliance/conformance/families/settings_load.py
"""Generic settings loading owns counts, attributed errors and cache effects."""

import json
from collections.abc import Mapping
from pathlib import Path
from typing import Any

from ..coverage import CoveragePredicate, FamilyCoveragePolicy

SYMBOLS = (
    "load_settings_sync",
    "load_settings_async",
    "load_batch_sync",
    "load_batch_async",
    "is_cached",
    "invalidate",
    "clear_cache",
    "cache_size",
    "cache_keys",
    "cache_stats",
    "reset_cache_stats",
)
OPERATIONS = (
    None,
    "load_settings_sync",
    "load_settings_async",
    "load_batch_sync",
    "load_batch_async",
    "is_cached",
    "loadSettingsSync",
    "loadSettingsAsync",
    "loadBatchSync",
    "loadBatchAsync",
    "isCached",
    "settings_load_sync",
    "settings_load_async_blocking",
    "settings_load_batch_sync",
    "settings_load_batch_async_blocking",
    "settings_is_cached",
    "invalidate",
    "clear_cache",
    "cache_size",
    "cache_keys",
    "cache_stats",
    "reset_cache_stats",
    "invalidateSettings",
    "clearSettingsCache",
    "settingsCacheSize",
    "settingsCacheKeys",
    "getSettingsCacheStats",
    "resetSettingsCacheStats",
    "settings_invalidate",
    "settings_clear_cache",
    "settings_cache_size",
    "settings_cache_keys",
    "settings_cache_stats",
    "settings_reset_cache_stats",
    "settings_cache_clear",
    "reset_settings_cache_stats",
    "yaml_ops_clear_cache",
)


def _relative(value: Any) -> bool:
    """Permit only portable, contained fixture paths before filesystem writes."""
    return (
        isinstance(value, str)
        and bool(value)
        and not any(char in value for char in "\\:")
        and all(part not in {"", ".", ".."} for part in value.split("/"))
    )


def _observed(value: Mapping[str, Any]) -> bool:
    """Require all loader outcomes, explicit cache effects and final input bytes."""
    if set(value) != {
        "sync",
        "async",
        "batchSync",
        "batchAsync",
        "files",
    } or not isinstance(value["files"], dict):
        return False
    if not all(
        _relative(path) and isinstance(content, str)
        for path, content in value["files"].items()
    ):
        return False
    for operation in ("sync", "async", "batchSync", "batchAsync"):
        result = value[operation]
        if not isinstance(result, dict) or set(result) != {
            "count",
            "error",
            "cached",
            "afterClear",
            "cacheState",
        }:
            return False
        state = result["cacheState"]
        if not isinstance(result["cached"], list) or not all(
            type(item) is bool for item in result["cached"]
        ):
            return False
        if not isinstance(state, dict) or set(state) != {
            "keys",
            "size",
            "invalidated",
            "invalidatedAgain",
            "afterInvalidate",
            "stats",
            "resetStats",
        }:
            return False
        if (
            not isinstance(state["keys"], list)
            or not all(isinstance(key, str) for key in state["keys"])
            or type(state["size"]) is not int
            or state["size"] != len(state["keys"])
            or state["afterInvalidate"] != 0
        ):
            return False
        if state["invalidated"] != result["cached"] or state["invalidatedAgain"] != [
            False
        ] * len(result["cached"]):
            return False
        for field in ("stats", "resetStats"):
            stats = state[field]
            if (
                not isinstance(stats, dict)
                or set(stats) != {"hits", "misses", "hitRateZero", "size", "bounded"}
                or stats["bounded"] is not True
                or any(
                    type(stats[k]) is not int or stats[k] < 0
                    for k in ("hits", "misses", "size")
                )
                or type(stats["hitRateZero"]) is not bool
            ):
                return False
        error = result["error"]
        if error is None:
            if type(result["count"]) is not int or result["count"] < 0:
                return False
        elif not (
            result["count"] is None
            and isinstance(error, dict)
            and set(error) == {"kind", "path"}
            and error["kind"] in {"io", "yaml-parse"}
            and _relative(error["path"])
        ):
            return False
        if not isinstance(result["cached"], list) or not all(
            type(item) is bool for item in result["cached"]
        ):
            return False
        if (
            not isinstance(result["afterClear"], list)
            or len(result["afterClear"]) != len(result["cached"])
            or any(item is not False for item in result["afterClear"])
        ):
            return False
    return True


def validate_settings_load_pack(
    document: Mapping[str, Any], root: Path
) -> tuple[Path, ...]:
    """Validate input-only requests and authored observation structure independently."""
    fixture_root = (root / document["fixtureRoot"]).resolve()
    paths = []
    for case in document["scenarios"]:
        reference = case["input"].get("fixtureRef")
        if (
            case["action"] != "settings-load.execute"
            or case["input"] != {"fixtureRef": reference}
            or case["fixtureRefs"] != [reference]
        ):
            raise ValueError("settings scenario must declare its sole input fixture")
        path = (fixture_root / document["fixtures"][reference]).resolve()
        if not path.is_relative_to(fixture_root):
            raise ValueError("settings fixture escapes fixture root")
        fixture = json.loads(path.read_text(encoding="utf-8"))
        if (
            not isinstance(fixture, dict)
            or set(fixture) != {"files", "request"}
            or not isinstance(fixture["files"], dict)
        ):
            raise ValueError("malformed settings fixture")
        request = fixture["request"]
        if (
            not isinstance(request, dict)
            or set(request) != {"single", "batch"}
            or not _relative(request["single"])
            or not isinstance(request["batch"], list)
            or not all(_relative(item) for item in request["batch"])
        ):
            raise ValueError("settings request needs contained relative paths")
        if not all(
            _relative(path) and isinstance(content, str)
            for path, content in fixture["files"].items()
        ) or not _observed(case["expected"]):
            raise ValueError("settings files or observation malformed")
        paths.append(path)
    return tuple(paths)


def settings_load_coverage_policy() -> FamilyCoveragePolicy:
    """Credit executed loaders and cache lifecycle operations, not YAML class methods."""
    return FamilyCoveragePolicy(
        "settings-load",
        (
            CoveragePredicate(
                id="settings-load-observed",
                capability_id="settings-load.execute",
                action="settings-load.execute",
                observation_family="settings-load",
                rust_symbols=SYMBOLS,
                matches=_observed,
                runtime_operations=OPERATIONS,
            ),
        ),
    )


YAML_SYMBOLS = (
    "YamlOperations",
    "new",
    "parse_yaml",
    "dump_yaml",
    "load_yaml_file",
    "save_yaml_file",
    "get_setting",
    "get_string_value",
    "get_vec_value",
    "get_hashmap_value",
    "set_setting",
    "yaml_cache_stats",
    "YamlCacheStats",
)
YAML_OPERATIONS = (
    None,
    "__init__",
    "parse_yaml",
    "dump_yaml",
    "load_yaml_file",
    "save_yaml_file",
    "get_setting",
    "get_string_value",
    "get_vec_value",
    "get_hashmap_value",
    "set_setting",
    "clear_cache",
    "get_cache_stats",
    "yaml_ops_new",
    "yaml_ops_parse",
    "yaml_ops_dump",
    "yaml_ops_load_file",
    "yaml_ops_save_file",
    "yaml_ops_get_setting_value",
    "yaml_ops_get_string",
    "yaml_ops_get_vec",
    "yaml_ops_set_string_setting",
    "yaml_ops_set_integer_setting",
    "yaml_ops_set_bool_setting",
    "yaml_ops_set_vec_setting",
    "yaml_ops_has_document",
    "yaml_ops_cache_size",
    "yaml_ops_cache_stats",
    "yamlParse",
    "yamlStringify",
    "yamlLoadFile",
    "yamlSaveFile",
    "yamlGetValue",
    "yamlGetStringValue",
    "yamlGetVecValue",
    "yamlGetHashmapValue",
    "yamlSetSetting",
)


def _yaml_observed(value: Mapping[str, Any]) -> bool:
    """Require typed values, actual saved bytes and a complete cache transition."""
    if set(value) != {"before", "after", "persisted", "files", "cache"}:
        return False
    before, after, persisted, files, cache = (
        value[key] for key in ("before", "after", "persisted", "files", "cache")
    )
    return (
        isinstance(before, dict)
        and set(before) == {"name", "missing", "items", "mapping"}
        and isinstance(before["name"], str)
        and isinstance(before["missing"], str)
        and isinstance(before["items"], list)
        and all(isinstance(item, str) for item in before["items"])
        and isinstance(before["mapping"], dict)
        and all(isinstance(item, str) for item in before["mapping"].values())
        and all(
            isinstance(item, dict)
            and set(item) == {"name", "ready", "size", "values"}
            and isinstance(item["name"], str)
            and type(item["ready"]) is bool
            and type(item["size"]) is int
            and isinstance(item["values"], list)
            and all(isinstance(entry, str) for entry in item["values"])
            for item in (after, persisted)
        )
        and isinstance(files, dict)
        and set(files) == {"saved.yaml"}
        and isinstance(files["saved.yaml"], str)
        and isinstance(cache, dict)
        and set(cache) == {"hits", "misses", "size", "afterClear"}
        and all(type(entry) is int and entry >= 0 for entry in cache.values())
    )


def validate_settings_yaml_pack(
    document: Mapping[str, Any], root: Path
) -> tuple[Path, ...]:
    """Validate input-only YAML fixtures independently of authored observations."""
    fixture_root = (root / document["fixtureRoot"]).resolve()
    paths = []
    for case in document["scenarios"]:
        reference = case["input"].get("fixtureRef")
        if (
            case["action"] != "settings-yaml.execute"
            or case["input"] != {"fixtureRef": reference}
            or case["fixtureRefs"] != [reference]
        ):
            raise ValueError("YAML scenario must declare its sole input fixture")
        path = (fixture_root / document["fixtures"][reference]).resolve()
        if not path.is_relative_to(fixture_root):
            raise ValueError("YAML fixture escapes root")
        fixture = json.loads(path.read_text(encoding="utf-8"))
        if (
            set(fixture) != {"kind", "content", "updates", "updateOrder"}
            or fixture["kind"] != "yaml"
            or not isinstance(fixture["content"], str)
            or not _yaml_observed(case["expected"])
        ):
            raise ValueError("malformed YAML fixture or observation")
        updates = fixture["updates"]
        if (
            not isinstance(updates, dict)
            or set(updates) != {"name", "ready", "size", "values"}
            or not isinstance(updates["name"], str)
            or type(updates["ready"]) is not bool
            or type(updates["size"]) is not int
            or not isinstance(updates["values"], list)
            or not all(isinstance(item, str) for item in updates["values"])
        ):
            raise ValueError("malformed YAML typed updates")
        if fixture["updateOrder"] != ["name", "ready", "size", "values"]:
            raise ValueError("YAML updates require explicit common order")
        paths.append(path)
    return tuple(paths)


def settings_yaml_coverage_policy() -> FamilyCoveragePolicy:
    """Credit only public methods executed by the typed YAML lifecycle adapters."""
    return FamilyCoveragePolicy(
        "settings-yaml",
        (
            CoveragePredicate(
                id="settings-yaml-observed",
                capability_id="settings-yaml.execute",
                action="settings-yaml.execute",
                observation_family="settings-yaml",
                rust_symbols=YAML_SYMBOLS,
                matches=_yaml_observed,
                runtime_operations=YAML_OPERATIONS,
            ),
        ),
    )


def _yaml_batch_observed(value: Mapping[str, Any]) -> bool:
    """Retain ordered entry arrays and typed batch/mapping results."""
    return (
        set(value) == {"before", "ordered", "vectors", "after"}
        and all(
            isinstance(value[key], dict)
            and all(isinstance(item, str) for item in value[key].values())
            for key in ("before", "after")
        )
        and isinstance(value["ordered"], list)
        and all(
            isinstance(pair, list)
            and len(pair) == 2
            and all(isinstance(item, str) for item in pair)
            for pair in value["ordered"]
        )
        and isinstance(value["vectors"], dict)
        and all(
            isinstance(items, list) and all(isinstance(item, str) for item in items)
            for items in value["vectors"].values()
        )
    )


def validate_settings_yaml_batch_pack(
    document: Mapping[str, Any], root: Path
) -> tuple[Path, ...]:
    """Reject malformed or expectation-bearing batch YAML fixtures."""
    fixture_root = (root / document["fixtureRoot"]).resolve()
    paths = []
    for case in document["scenarios"]:
        reference = case["input"].get("fixtureRef")
        if (
            case["action"] != "settings-yaml-batch.execute"
            or case["input"] != {"fixtureRef": reference}
            or case["fixtureRefs"] != [reference]
        ):
            raise ValueError("YAML batch fixture must be solely declared")
        path = (fixture_root / document["fixtures"][reference]).resolve()
        if not path.is_relative_to(fixture_root):
            raise ValueError("YAML batch fixture escapes root")
        fixture = json.loads(path.read_text(encoding="utf-8"))
        if (
            set(fixture) != {"kind", "content", "keys", "updates"}
            or fixture["kind"] != "yaml-batch"
            or not isinstance(fixture["content"], str)
            or not isinstance(fixture["keys"], list)
            or not all(isinstance(key, str) for key in fixture["keys"])
            or not isinstance(fixture["updates"], dict)
            or not all(isinstance(value, str) for value in fixture["updates"].values())
            or not _yaml_batch_observed(case["expected"])
        ):
            raise ValueError("malformed YAML batch fixture or observation")
        paths.append(path)
    return tuple(paths)


def settings_yaml_batch_coverage_policy() -> FamilyCoveragePolicy:
    """Credit only the batch and ordered-map public operations actually invoked."""
    symbols = (
        "get_settings_batch",
        "set_settings_batch",
        "get_indexmap_value",
        "get_hashmap_vec_value",
    )
    return FamilyCoveragePolicy(
        "settings-yaml-batch",
        (
            CoveragePredicate(
                id="settings-yaml-batch-observed",
                capability_id="settings-yaml-batch.execute",
                action="settings-yaml-batch.execute",
                observation_family="settings-yaml-batch",
                rust_symbols=symbols,
                matches=_yaml_batch_observed,
                runtime_operations=(
                    None,
                    "yamlGetSettingsBatch",
                    "yamlSetSettingsBatch",
                    "yamlGetIndexmapValue",
                    "yamlGetHashmapVecValue",
                ),
            ),
        ),
    )

File: business-logic/classic-scanlog-core/tests/semantic_conformance/settings_load.rs
//! Generic settings loading through public sync/async APIs on the shared runtime.

use super::{RunnerResult, invalid, strings, text};
use classic_settings_core as settings;
use classic_shared_core::get_runtime;
use serde_json::{Value, json};
use std::{
    fs,
    path::{Path, PathBuf},
};

/// Ensure global cache cleanup on every scenario exit, including native errors.
struct CacheReset;
impl Drop for CacheReset {
    fn drop(&mut self) {
        settings::clear_cache();
    }
}

/// Clear scenario-owned global YAML cache state even when a native operation fails.
struct YamlCacheReset;
impl Drop for YamlCacheReset {
    fn drop(&mut self) {
        settings::clear_global_yaml_cache();
    }
}

/// Contain authored destinations before creating scenario-owned files.
fn owned(root: &Path, relative: &str) -> RunnerResult<PathBuf> {
    if relative.is_empty()
        || relative.contains(['\\', ':'])
        || relative
            .split('/')
            .any(|part| matches!(part, "" | "." | ".."))
    {
        return Err(invalid("settings fixture requires a contained relative path").into());
    }
    Ok(root.join(relative))
}

/// Preserve typed core failures and their attributed path without OS-specific prose.
fn domain_error(error: settings::SettingsError, root: &Path) -> RunnerResult<Value> {
    let (kind, path) = match error {
        settings::SettingsError::IoError { path, .. } => ("io", path),
        settings::SettingsError::YamlParseError { source, .. } => (
            "yaml-parse",
            source
                .path()
                .ok_or_else(|| invalid("expected file-backed parse error"))?
                .clone(),
        ),
        other => return Err(other.into()),
    };
    Ok(json!({"kind": kind, "path": path.strip_prefix(root)?.to_string_lossy().replace('\\', "/")}))
}

/// Inventory every resulting file so unexpected writes cannot hide behind input paths.
pub(super) fn inventory(
    root: &Path,
    directory: &Path,
    files: &mut serde_json::Map<String, Value>,
) -> RunnerResult<()> {
    for entry in fs::read_dir(directory)? {
        let path = entry?.path();
        if path.is_dir() {
            inventory(root, &path, files)?;
        } else {
            files.insert(
                path.strip_prefix(root)?
                    .to_string_lossy()
                    .replace('\\', "/"),
                Value::String(fs::read_to_string(path)?),
            );
        }
    }
    Ok(())
}

/// Observe document/file counts, cache effects and attributed sync/async failures.
pub(super) fn execute(fixture: &Value) -> RunnerResult<Value> {
    if fixture["kind"] == "yaml-batch" {
        return execute_yaml_batch(fixture);
    }
    if fixture["kind"] == "yaml" {
        return execute_yaml(fixture);
    }
    let temporary = tempfile::tempdir()?;
    let root = temporary.path();
    let _reset = CacheReset;
    for (relative, content) in fixture["files"]
        .as_object()
        .ok_or_else(|| invalid("expected files object"))?
    {
        let path = owned(root, relative)?;
        fs::create_dir_all(path.parent().ok_or_else(|| invalid("missing parent"))?)?;
        fs::write(path, text(content)?.as_bytes())?;
    }
    settings::reset_cache_stats();
    let mut observed = serde_json::Map::new();
    for operation in ["sync", "async", "batchSync", "batchAsync"] {
        settings::clear_cache();
        let batch = operation.starts_with("batch");
        let relatives = if batch {
            strings(&fixture["request"]["batch"])?
        } else {
            vec![text(&fixture["request"]["single"])?]
        };
        let paths = relatives
            .iter()
            .map(|relative| owned(root, relative))
            .collect::<RunnerResult<Vec<_>>>()?;
        let refs: Vec<_> = paths.iter().map(PathBuf::as_path).collect();
        let keys = if batch {
            paths
                .iter()
                .map(|path| path.to_string_lossy().to_string())
                .collect()
        } else {
            vec!["conformance.single".to_string()]
        };
        let result = match operation {
            "sync" => settings::load_settings_sync(&keys[0], &paths[0]).map(|docs| docs.len()),
            "async" => get_runtime()
                .block_on(settings::load_settings_async(&keys[0], &paths[0]))
                .map(|docs| docs.len()),
            "batchSync" => settings::load_batch_sync(&refs),
            _ => get_runtime().block_on(settings::load_batch_async(&refs)),
        };
        let (count, error) = match result {
            Ok(count) => (json!(count), Value::Null),
            Err(error) => (Value::Null, domain_error(error, root)?),
        };
        let cached: Vec<_> = keys.iter().map(|key| settings::is_cached(key)).collect();
        let mut normalized_keys: Vec<_> = settings::cache_keys()
            .iter()
            .map(|key| {
                if batch {
                    Path::new(key)
                        .strip_prefix(root)
                        .unwrap()
                        .to_string_lossy()
                        .replace('\\', "/")
                } else {
                    key.clone()
                }
            })
            .collect();
        normalized_keys.sort();
        let size = settings::cache_size();
        let stats = cache_statistics();
        let invalidated: Vec<_> = keys.iter().map(|key| settings::invalidate(key)).collect();
        let invalidated_again: Vec<_> = keys.iter().map(|key| settings::invalidate(key)).collect();
        let after_invalidate = settings::cache_size();
        settings::reset_cache_stats();
        let reset_stats = cache_statistics();
        let cache_state = json!({"keys": normalized_keys, "size": size, "stats": stats, "invalidated": invalidated, "invalidatedAgain": invalidated_again, "afterInvalidate": after_invalidate, "resetStats": reset_stats});
        // Refill successfully loaded entries so clearing is observed independently of invalidation.
        for ((key, path), present) in keys.iter().zip(&paths).zip(&cached) {
            if *present {
                settings::load_settings_sync(key, path)?;
            }
        }
        settings::clear_cache();
        let after_clear: Vec<_> = keys.iter().map(|key| settings::is_cached(key)).collect();
        observed.insert(
            operation.to_string(),
            json!({"count": count, "error": error, "cached": cached, "afterClear": after_clear, "cacheState": cache_state}),
        );
    }
    let mut files = serde_json::Map::new();
    inventory(root, root, &mut files)?;
    observed.insert("files".to_string(), Value::Object(files));
    Ok(Value::Object(observed))
}

/// Project stable cache counters while retaining whether storage is bounded.
fn cache_statistics() -> Value {
    let stats = settings::cache_stats();
    json!({"hits": stats.hits, "misses": stats.misses, "hitRateZero": stats.hit_rate == 0.0, "size": stats.size, "bounded": stats.capacity > 0})
}

/// Project authored JSON-compatible YAML values without hiding typed scalar changes.
pub(super) fn yaml_json(value: &settings::Yaml) -> RunnerResult<Value> {
    use settings::Yaml;
    Ok(match value {
        Yaml::String(value) => json!(value),
        Yaml::Integer(value) => json!(value),
        Yaml::Boolean(value) => json!(value),
        Yaml::Null | Yaml::BadValue => Value::Null,
        Yaml::Array(values) => {
            Value::Array(values.iter().map(yaml_json).collect::<RunnerResult<_>>()?)
        }
        Yaml::Hash(values) => Value::Object(
            values
                .iter()
                .map(|(key, value)| {
                    Ok((
                        key.as_str()
                            .ok_or_else(|| invalid("expected string YAML key"))?
                            .to_string(),
                        yaml_json(value)?,
                    ))
                })
                .collect::<RunnerResult<_>>()?,
        ),
        _ => return Err(invalid("unexpected non-JSON YAML value").into()),
    })
}

/// Observe YAML handles through parsing, typed access, persisted mutation and cache reuse.
fn execute_yaml(fixture: &Value) -> RunnerResult<Value> {
    let temporary = tempfile::tempdir()?;
    let _reset = YamlCacheReset;
    let ops = settings::YamlOperations::new();
    ops.clear_cache();
    let mut doc = ops.parse_yaml(&text(&fixture["content"])?)?;
    let before = json!({"name": ops.get_string_value(&doc, "name", "fallback"), "missing": ops.get_string_value(&doc, "absent", "fallback"), "items": ops.get_vec_value(&doc, "items"), "mapping": ops.get_hashmap_value(&doc, "mapping")});
    let updates = fixture["updates"]
        .as_object()
        .ok_or_else(|| invalid("expected updates"))?;
    for key in strings(&fixture["updateOrder"])? {
        let value = &updates[&key];
        let native = match value {
            Value::String(value) => settings::Yaml::String(value.clone()),
            Value::Bool(value) => settings::Yaml::Boolean(*value),
            Value::Number(value) => {
                settings::Yaml::Integer(value.as_i64().ok_or_else(|| invalid("expected integer"))?)
            }
            Value::Array(values) => settings::Yaml::Array(
                values
                    .iter()
                    .map(|value| Ok(settings::Yaml::String(text(value)?.to_string())))
                    .collect::<RunnerResult<_>>()?,
            ),
            _ => return Err(invalid("unsupported fixture update").into()),
        };
        doc = ops.set_setting(&doc, &key, native)?;
    }
    let round_trip = ops.parse_yaml(&ops.dump_yaml(&doc)?)?;
    let after = updates
        .keys()
        .map(|key| {
            Ok((
                key.clone(),
                yaml_json(
                    &ops.get_setting(&round_trip, key)
                        .unwrap_or(settings::Yaml::Null),
                )?,
            ))
        })
        .collect::<RunnerResult<serde_json::Map<_, _>>>()?;
    let path = temporary.path().join("saved.yaml");
    ops.save_yaml_file(&path, &doc)?;
    let initial = settings::yaml_cache_stats();
    let loaded = ops.load_yaml_file(&path)?;
    ops.load_yaml_file(&path)?;
    let stats = settings::yaml_cache_stats();
    let persisted = updates
        .keys()
        .map(|key| {
            Ok((
                key.clone(),
                yaml_json(
                    &ops.get_setting(&loaded, key)
                        .unwrap_or(settings::Yaml::Null),
                )?,
            ))
        })
        .collect::<RunnerResult<serde_json::Map<_, _>>>()?;
    let mut files = serde_json::Map::new();
    inventory(temporary.path(), temporary.path(), &mut files)?;
    ops.clear_cache();
    ops.load_yaml_file(&path)?;
    if settings::yaml_cache_stats().size != 1 {
        return Err(invalid("global clear fixture did not populate YAML cache").into());
    }
    settings::clear_global_yaml_cache();
    Ok(
        json!({"before": before, "after": after, "persisted": persisted, "files": files, "cache": {"hits": stats.hits - initial.hits, "misses": stats.misses - initial.misses, "size": stats.size, "afterClear": settings::yaml_cache_stats().size}}),
    )
}

/// Observe batch access and order-preserving mappings through Rust's public YAML API.
fn execute_yaml_batch(fixture: &Value) -> RunnerResult<Value> {
    let ops = settings::YamlOperations::new();
    let doc = ops.parse_yaml(&text(&fixture["content"])?)?;
    let keys = strings(&fixture["keys"])?;
    let key_refs: Vec<_> = keys.iter().map(String::as_str).collect();
    let project =
        |values: std::collections::HashMap<String, settings::Yaml>| -> RunnerResult<Value> {
            Ok(Value::Object(
                values
                    .iter()
                    .map(|(key, value)| Ok((key.clone(), yaml_json(value)?)))
                    .collect::<RunnerResult<_>>()?,
            ))
        };
    let before = project(ops.get_settings_batch(&doc, &key_refs))?;
    let ordered: Vec<_> = ops
        .get_indexmap_value(&doc, "mapping")
        .into_iter()
        .collect();
    let vectors = ops.get_hashmap_vec_value(&doc, "vectors");
    let updates = fixture["updates"]
        .as_object()
        .ok_or_else(|| invalid("expected batch updates"))?;
    let values: Vec<_> = updates
        .iter()
        .map(|(key, value)| {
            Ok((
                key.as_str(),
                settings::Yaml::String(text(value)?.to_string()),
            ))
        })
        .collect::<RunnerResult<_>>()?;
    let updated = ops.set_settings_batch(&doc, &values)?;
    Ok(
        json!({"before": before, "ordered": ordered, "vectors": vectors, "after": project(ops.get_settings_batch(&updated, &key_refs))?}),
    )
}
